File: source/analysis/pre1790/analyze_by_year.jl
using CSV, DataFrames, DataFramesMeta, Gadfly, XLSX, Dates, Compose, Plots

function clean_df(df) 
    df[:, :dollars] = replace(df[:, :dollars], "NaN" => missing) # replace "NaN" with missing
    df[:, :cents] = replace(df[:, :cents], "NaN" => missing) # replace "NaN" with missing
    df[:, :dollars] = replace(df[:, :dollars], missing => 0)  # replace missing with 0 
    df[:, :cents] = replace(df[:, :cents], missing => 0)  # replace missing with 0 
    df 
end 

function group_by_year(df) 
    gdf = @by(df, :year,
        :total_amt = sum(:dollars) + sum(:cents)  # calculate total amount of debt 
    )
    sort!(gdf) # sort by year 
    replace!(gdf.year, "missing" => "no year") # change missing values to string to allow for plotting
    return gdf 
end

function group_by_year_month(df)
    gdf = @by(df, [:year, :month],
        :total_amt = sum(:dollars) + sum(:cents)  # calculate total amount of debt 
    )
    
    return gdf 
end

#plot debt redeemed per year saved as svg 
function plot_debt(df::DataFrame, state, type)
    set_default_plot_size(40cm, 22.5cm)

    if type == "year"
        p_debt_date = Gadfly.plot(
            df,
            x=:year,
            y=:total_amt,
            Geom.bar,
            Guide.xlabel("Year"),
            Guide.ylabel("Total Debt (in dollars)"),
            Guide.title(uppercase(state)*" Debt Redeemed Per Year"),
            Gadfly.Theme(background_color = "white")
        )
        img = SVG("output/analysis/pre1790/debt_per_year/year/" * uppercase(state) * "_debt_redeemed_per_year.svg", 40cm, 22.5cm) 
        draw(img, p_debt_date)  
    elseif type == "year_month"
        p_debt_date = Gadfly.plot(
            df,
            x=:year_month,
            y=:total_amt,
            Geom.bar,
            Guide.xlabel("Year"),
            Guide.ylabel("Total Debt (in dollars)"),
            Guide.title(uppercase(state)*" Debt Redeemed Per Year/Month"),
            Gadfly.Theme(background_color = "white")
        )

        img = SVG("output/analysis/pre1790/debt_per_year/year_month/" * uppercase(state) * "_debt_redeemed_per_year_month.svg", 40cm, 22.5cm) 
        draw(img, p_debt_date)  
    end

    println(state)

end 

function group_post1795(state_df) 
    # filtering dataframes using multiple conditions 
    state_df_post1795 = @subset(state_df, :year_month .> DateTime(1795)) # filter out years before 1795
    state_df_post1795.total_amt = replace(state_df_post1795.total_amt, missing => 0) # replace missing values with 0
    state_df_sum = sum(state_df_post1795.total_amt) # sum total debt redeemed after 1795

    # remove rows where year is greater than 1795 
    state_df = @subset(state_df, :year_month .< DateTime(1795))
    
    # change datetime object to string 
    state_df.year_month = Dates.format.(state_df.year_month, "yyyy-mm")
    
    push!(state_df, [state_df_sum, "post-1795"], promote=true) # add total debt redeemed after 1795 to dataframe
    return state_df
end 

function handle_missing_info(state_df)
    missing_info = @rsubset(state_df, ismissing(:year) || ismissing(:month)) # filter out rows with missing year or month\
    missing_info_sum = 0
    if nrow(missing_info) > 0
        missing_info_sum = sum(missing_info.dollars + missing_info.cents) # sum total debt redeemed with missing year or month
    end 
    return missing_info_sum
end 

# import cd_info 
cd_info = DataFrame(CSV.File("output/derived/pre1790/cd_info.csv"))
# store total amount and years of all states in a new dataframe 
all_states = DataFrame([[], []], [:year_month, :total_amt])

# loop through cd_info per state 
for i in 1:nrow(cd_info)
    state_df = DataFrame() # create new dataframe for each state
    state_row = cd_info[i, :]
    state_excel = XLSX.readxlsx(state_row[:file_path])
    state_sheet = state_excel["Sheet1"]
    
    # get year columns --> merge years columns into one column 
    year_col_indexes = split(state_row[:year_col], ",")
    for year_col_index in year_col_indexes
        year_col_range = year_col_index*string(state_row[:first_row])*":"*year_col_index*string(state_row[:last_row])
        year_col = state_sheet[year_col_range]
        
        if nrow(state_df) == 0 # if state_df is empty, create year column 
            state_df.year = vec(year_col)
        else 
            state_df.year .= coalesce(state_df.year, vec(year_col)) # append year col to state dataframe
        end 
    end
    
    # get month columns --> merge month columns into one column
    month_col_indexes = split(state_row[:month_col], ",")
    for month_col_index in month_col_indexes
        month_col_range = month_col_index*string(state_row[:first_row])*":"*month_col_index*string(state_row[:last_row])
        month_col = state_sheet[month_col_range]

        if !("month" in names(state_df)) 
            state_df.month = vec(month_col)
        else
            state_df.month .= coalesce(state_df.month, vec(month_col)) # append month col to state dataframe
        end 
    end
    
    # add dollar amount column to states dataframe 
    dollar_indexes = split(state_row[:dollars_col], ",") 
    for dollar_index in dollar_indexes
        dollar_col_range = dollar_index*string(state_row[:first_row])*":"*dollar_index*string(state_row[:last_row])
        dollar_col = state_sheet[dollar_col_range]
        state_df.dollars = vec(dollar_col)
    end

    # add cents amount column to states dataframe 
    if !ismissing(state_row[:cents_col]) # handle excel spreadsheets with no cents column 
        cents_indexes = split(state_row[:cents_col], ",")
        for cents_index in cents_indexes
            cents_col_range = cents_index*string(state_row[:first_row])*":"*cents_index*string(state_row[:last_row])
            cents_col = state_sheet[cents_col_range]
            state_df.cents = vec(cents_col) ./ 100 # convert cents to decimal
        end 
    else 
        state_df.cents = zeros(nrow(state_df)) # create column of zeros to allow for summing
    end

    state_df = clean_df(state_df) # clean dataframe to removeq missing values from :dollars and :cents columns
    missing_info_sum = handle_missing_info(state_df) # get total value of rows with missing info 
    dropmissing!(state_df, [:year, :month]) # drop rows with missing values in year, month columns

    state_df.year = string.(state_df.year)
    state_df.month = string.(state_df.month)

    # add state label 
    state_df.state = fill(state_row[:state], nrow(state_df))

    state_df_clean = clean_df(state_df) # clean table to remove missing values
    state_gdf = group_by_year_month(state_df_clean) # group by year and month and sum debt 

    # create new column that merges year and month 
    state_gdf.year_month = string.(state_gdf[:, :year], "-", state_gdf[:, :month])

    # convert year_month column to datetime object 
    state_gdf.year_month = DateTime.(state_gdf.year_month, DateFormat("yyyy-mm"))
    sort!(state_gdf, :year_month) # sort by year_month column   

    # remove year and month columns 
    select!(state_gdf, Not([:year, :month]))

    all_states = vcat(all_states, state_gdf) # append state dataframe to all_states dataframe

    state_gdf = group_post1795(state_gdf) # filter out years before 1795
    push!(state_gdf, [missing_info_sum, "missing info"], promote=true) # add total debt redeemed with missing info to dataframe

    plot_debt(state_gdf, state_row[:state], "year_month") # plot debt redeemed per year saved as svg
    #plot_debt(state_gdf, state_row[:state], "year") # plot debt redeemed per year saved as svg

    println(last(state_gdf, 5)) 
end 


# plot united states 
all_states = @by(all_states, :year_month, :total_amt = sum(:total_amt)) # group by year and month and sum debt
all_states = all_states[!, [:total_amt, :year_month]] # swap columns
sort!(all_states, :year_month) # sort by year_month column
all_states = group_post1795(all_states) # filter out years before 1795

# import pre-1790 debt data 
# get unique dates 
# group by year - sum debt 
# plot debt redeemed per year saved as svg 

pre1790 = DataFrame(CSV.File("output/derived/pre1790/pre1790_cleaned.csv"))
pre1790.year = pre1790[:, "date of the certificate | year"]

# fix cents column in pre1790_cleaned.csv 
pre1790[:, "amount | 90th"] = getindex.(split.(pre1790[:, "amount | 90th"], "."), 1)
pre1790[:, "amount | 90th"] = replace.(pre1790[:, "amount | 90th"], "/" => "")
pre1790.cents = parse.(Float64, pre1790[:, "amount | 90th"]) ./ 100

pre1790.total_amt = pre1790[:, "amount | dollars"] + pre1790[:, "cents"]
pre1790.dollars = pre1790[:, "amount | dollars"]

pre1790.cents = ifelse.(pre1790.cents .>= 100, 0, pre1790.cents)

#clean 
pre1790_clean = clean_df(pre1790)

#group by year and sum 
pre1790_clean.year = coalesce.(pre1790_clean.year, 0)
pre1790_clean.year = Int.(pre1790_clean.year)
pre1790_clean.year = string.(pre1790_clean.year)
pre1790_gdf = group_by_year(pre1790_clean)
pre1790_gdf[1, "year"] = "no year"
sort!(pre1790_gdf)

pre1790_gdf.total_amt = pre1790_gdf.total_amt ./ 1e6 # convert to millions

#plot 
p_bar = bar(pre1790_gdf.year, pre1790_gdf.total_amt, 
    title = "Debt Certificate Total Per Year",
    xlabel = "Year",
    ylabel = "Amount (in million dollars)",
    legend = false,
    background_color = "white",
    yaxis=[0, 30],
    top_margin=5mm,
)

# round to 2 decimal places
pre1790_gdf.total_amt = round.(pre1790_gdf.total_amt, digits=2)
annotate!(pre1790_gdf.year, pre1790_gdf.total_amt, pre1790_gdf.total_amt, annotationfontsizes=8, annotationvalign=:bottom)

savefig(p_bar, "output/analysis/pre1790/debt_per_year/pre1790_debt_certificate_amts_per_year.svg")

#plot 
total_amt = sum(pre1790_gdf.total_amt)
pre1790_gdf.percent = pre1790_gdf.total_amt ./ total_amt

p_bar = bar(pre1790_gdf.year, pre1790_gdf.percent, 
    title = "Percent of Debt Certificate Total Per Year",
    xlabel = "Year",
    ylabel = "Amount (Percentage of Total)",
    legend = false,
    background_color = "white",
    yaxis=[0, 0.5],
    top_margin=5mm
)

# round to 2 decimal places
pre1790_gdf.percent = round.(pre1790_gdf.percent, digits=2)
annotate!(pre1790_gdf.year, pre1790_gdf.percent, pre1790_gdf.percent, annotationfontsizes=8, annotationvalign=:bottom)

savefig(p_bar, "output/analysis/pre1790/debt_per_year/pre1790_debt_certificate_percent_per_year.svg")

File: source/analysis/pre1790/analyze_by_year.py
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from openpyxl import load_workbook

INDIR_CD_INFO = Path("source/raw/post1790_cd")
INDIR_RAW = Path("source/raw/post1790_cd/orig")
INDIR_PRE1790 = Path("output/derived/prescrape/pre1790")
OUTDIR = Path("output/analysis/pre1790/debt_per_year")
OUTDIR.mkdir(parents=True, exist_ok=True)
(OUTDIR / "year_month").mkdir(parents=True, exist_ok=True)


def Main():
    PlotPre1790ByYear()
    PlotPre1790PctByYear()

    cd_info_path = INDIR_CD_INFO / "cd_info.csv"
    if not cd_info_path.exists():
        print(f"Skipping per-state plots: {cd_info_path} not found (Task 4 item)")
        return
    cd_info = pd.read_csv(cd_info_path)
    all_states_frames = []

    for _, row in cd_info.iterrows():
        state_df = LoadStateXlsx(row)
        if state_df is None or state_df.empty:
            continue
        state_df = CleanDf(state_df)
        missing_sum = HandleMissingInfo(state_df)
        state_df = state_df.dropna(subset=["year", "month"])
        state_df["year"] = state_df["year"].astype(str)
        state_df["month"] = state_df["month"].astype(str).str.zfill(2)
        state_df["year_month"] = state_df["year"] + "-" + state_df["month"]
        gdf = GroupByYearMonth(state_df)
        gdf = GroupPost1795(gdf, missing_sum)
        PlotDebt(gdf, row["state"])
        all_states_frames.append(gdf[["year_month", "total_amt"]].copy())

    if all_states_frames:
        all_states = pd.concat(all_states_frames, ignore_index=True)
        us_gdf = all_states.groupby("year_month", as_index=False)["total_amt"].sum()
        us_gdf = GroupPost1795(us_gdf, 0)
        PlotDebt(us_gdf, "US")


def LoadStateXlsx(row):
    file_path = Path(row["file_path"])
    if not file_path.exists():
        file_path = INDIR_RAW / file_path.name
    if not file_path.exists():
        return None
    try:
        wb = load_workbook(file_path, read_only=True, data_only=True)
        ws = wb["Sheet1"]
        first_row, last_row = int(row["first_row"]), int(row["last_row"])
        n = last_row - first_row + 1

        year = MergeCols(ws, str(row["year_col"]), first_row, last_row)
        month = MergeCols(ws, str(row["month_col"]), first_row, last_row)
        dollars = MergeCols(ws, str(row["dollars_col"]), first_row, last_row)
        cents_col = row.get("cents_col")
        if pd.notna(cents_col):
            cents_raw = MergeCols(ws, str(cents_col), first_row, last_row)
            cents = [float(v) / 100 if v is not None else 0.0 for v in cents_raw]
        else:
            cents = [0.0] * n
        wb.close()

        return pd.DataFrame({"year": year, "month": month, "dollars": dollars, "cents": cents})
    except Exception:
        return None


def MergeCols(ws, col_str, first_row, last_row):
    n = last_row - first_row + 1
    result = [None] * n
    for col_letter in col_str.split(","):
        col_letter = col_letter.strip()
        for i, r in enumerate(range(first_row, last_row + 1)):
            val = ws[f"{col_letter}{r}"].value
            if result[i] is None and val is not None:
                result[i] = val
    return result


def CleanDf(df):
    for col in ["dollars", "cents"]:
        df[col] = pd.to_numeric(df[col], errors="coerce").fillna(0)
    return df


def HandleMissingInfo(df):
    missing = df[df["year"].isna() | df["month"].isna()]
    return (missing["dollars"] + missing["cents"]).sum()


def GroupByYearMonth(df):
    df["total_amt"] = df["dollars"] + df["cents"]
    return df.groupby("year_month", as_index=False)["total_amt"].sum()


def GroupPost1795(gdf, missing_sum):
    gdf["_sort_key"] = pd.to_datetime(gdf["year_month"], format="%Y-%m", errors="coerce")
    post1795 = gdf[gdf["_sort_key"] > pd.Timestamp("1795-01-01")]
    pre1795 = gdf[gdf["_sort_key"] <= pd.Timestamp("1795-01-01")].drop(columns="_sort_key")
    post1795_sum = post1795["total_amt"].sum()
    extra_rows = []
    if post1795_sum > 0:
        extra_rows.append({"year_month": "post-1795", "total_amt": post1795_sum})
    if missing_sum > 0:
        extra_rows.append({"year_month": "missing info", "total_amt": missing_sum})
    if extra_rows:
        pre1795 = pd.concat([pre1795, pd.DataFrame(extra_rows)], ignore_index=True)
    return pre1795


def PlotDebt(gdf, state):
    fig, ax = plt.subplots(figsize=(14, 6))
    ax.bar(range(len(gdf)), gdf["total_amt"], tick_label=gdf["year_month"].tolist())
    ax.set_xlabel("Year/Month")
    ax.set_ylabel("Total Debt (in dollars)")
    ax.set_title(f"{state.upper()} Debt Redeemed Per Year/Month")
    plt.xticks(rotation=90, fontsize=7)
    plt.tight_layout()
    out = OUTDIR / "year_month" / f"{state.upper()}_debt_redeemed_per_year_month.png"
    plt.savefig(out, dpi=150)
    plt.close()


def PlotPre1790ByYear():
    pre1790 = pd.read_csv(INDIR_PRE1790 / "pre1790_cleaned.csv")
    pre1790["year"] = pre1790.get("date of the certificate | year")
    pre1790["dollars"] = pd.to_numeric(pre1790.get("amount | dollars"), errors="coerce").fillna(0)
    cents_raw = pre1790.get("amount | 90th", pd.Series(dtype=str))
    cents_raw = cents_raw.astype(str).str.split(".").str[0].str.replace("/", "")
    pre1790["cents"] = pd.to_numeric(cents_raw, errors="coerce").fillna(0) / 100
    pre1790.loc[pre1790["cents"] >= 100, "cents"] = 0
    pre1790["total_amt"] = pre1790["dollars"] + pre1790["cents"]

    gdf = pre1790.groupby("year", as_index=False)["total_amt"].sum()
    gdf["year"] = gdf["year"].fillna(0).astype(int).astype(str)
    gdf.loc[gdf["year"] == "0", "year"] = "no year"
    gdf = gdf.sort_values("year")
    gdf["total_amt_m"] = (gdf["total_amt"] / 1e6).round(2)

    fig, ax = plt.subplots(figsize=(12, 6))
    bars = ax.bar(range(len(gdf)), gdf["total_amt_m"], tick_label=gdf["year"].tolist())
    ax.bar_label(bars, labels=gdf["total_amt_m"].tolist(), fontsize=8, padding=2)
    ax.set_ylim(0, gdf["total_amt_m"].max() * 1.15)
    ax.set_xlabel("Year")
    ax.set_ylabel("Amount (in million dollars)")
    ax.set_title("Debt Certificate Total Per Year")
    plt.xticks(rotation=45, ha="right")
    plt.tight_layout()
    plt.savefig(OUTDIR / "pre1790_debt_certificate_amts_per_year.png", dpi=150)
    plt.close()


def PlotPre1790PctByYear():
    pre1790 = pd.read_csv(INDIR_PRE1790 / "pre1790_cleaned.csv")
    pre1790["year"] = pre1790.get("date of the certificate | year")
    pre1790["dollars"] = pd.to_numeric(pre1790.get("amount | dollars"), errors="coerce").fillna(0)
    cents_raw = pre1790.get("amount | 90th", pd.Series(dtype=str))
    cents_raw = cents_raw.astype(str).str.split(".").str[0].str.replace("/", "")
    pre1790["cents"] = pd.to_numeric(cents_raw, errors="coerce").fillna(0) / 100
    pre1790.loc[pre1790["cents"] >= 100, "cents"] = 0
    pre1790["total_amt"] = pre1790["dollars"] + pre1790["cents"]

    gdf = pre1790.groupby("year", as_index=False)["total_amt"].sum()
    gdf["year"] = gdf["year"].fillna(0).astype(int).astype(str)
    gdf.loc[gdf["year"] == "0", "year"] = "no year"
    gdf = gdf.sort_values("year")
    total = gdf["total_amt"].sum()
    gdf["percent"] = (gdf["total_amt"] / total).round(2)

    fig, ax = plt.subplots(figsize=(12, 6))
    bars = ax.bar(range(len(gdf)), gdf["percent"], tick_label=gdf["year"].tolist())
    ax.bar_label(bars, labels=gdf["percent"].tolist(), fontsize=8, padding=2)
    ax.set_ylim(0, gdf["percent"].max() * 1.15)
    ax.set_xlabel("Year")
    ax.set_ylabel("Amount (Percentage of Total)")
    ax.set_title("Percent of Debt Certificate Total Per Year")
    plt.xticks(rotation=45, ha="right")
    plt.tight_layout()
    plt.savefig(OUTDIR / "pre1790_debt_certificate_percent_per_year.png", dpi=150)
    plt.close()


if __name__ == "__main__":
    Main()
